src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -5127,100 +5128,113 @@ private void startLeaving()
+                            {
+                                NetworkTopologyStrategy strategy = (NetworkTopologyStrategy) keyspace.getReplicationStrategy();
+                                rf = strategy.getReplicationFactor(dc).allReplicas;
+                                numNodes = metadata.getTopology().getDatacenterEndpoints().get(dc).size();
+                            }
+                            else
+                            {
+                                numNodes = metadata.getAllEndpoints().size();
+                                rf = keyspace.getReplicationStrategy().getReplicationFactor().allReplicas;
+                            }
 
-                        if (numNodes <= rf)
-                            throw new UnsupportedOperationException("Not enough live nodes to maintain replication factor in keyspace "
-                                                                    + keyspaceName + " (RF = " + rf + ", N = " + numNodes + ")."
-                                                                    + " Perform a forceful decommission to ignore.");
+                            if (numNodes <= rf)
+                                throw new UnsupportedOperationException("Not enough live nodes to maintain replication factor in keyspace "
+                                                                        + keyspaceName + " (RF = " + rf + ", N = " + numNodes + ")."
+                                                                        + " Perform a forceful decommission to ignore.");
+                        }
+                        // TODO: do we care about fixing transient/full self-movements here? probably
+                        if (tokenMetadata.getPendingRanges(keyspaceName, FBUtilities.getBroadcastAddressAndPort()).size() > 0)
+                            throw new UnsupportedOperationException("data is currently moving to this node; unable to leave the ring");
                     }
-                    // TODO: do we care about fixing transient/full self-movements here? probably
-                    if (tokenMetadata.getPendingRanges(keyspaceName, FBUtilities.getBroadcastAddressAndPort()).size() > 0)
-                        throw new UnsupportedOperationException("data is currently moving to this node; unable to leave the ring");
                 }
-            }
 
-            startLeaving();
-            long timeout = Math.max(RING_DELAY_MILLIS, BatchlogManager.instance.getBatchlogTimeout());
-            setMode(Mode.LEAVING, "sleeping " + timeout + " ms for batch processing and pending range setup", true);
-            Thread.sleep(timeout);
+                startLeaving();
+                long timeout = Math.max(RING_DELAY_MILLIS, BatchlogManager.instance.getBatchlogTimeout());
+                setMode(Mode.LEAVING, "sleeping " + timeout + " ms for batch processing and pending range setup", true);
+                Thread.sleep(timeout);
```

Comment:
maybe it is good time to sleep uninterruptibly? Look into Uninterruptibles in google guava